File: 2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python3
#
# [2] Add Two Numbers
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy_head = ListNode(0)
        tmp = dummy_head
        carry = 0
        while (l1 is not None) or (l2 is not None):
            sum_ = carry

            if l1 is not None:
                sum_ += l1.val
                l1 = l1.next
            if l2 is not None:
                sum_ += l2.val 
                l2 = l2.next
            carry = sum_ // 10
            sum_ = sum_ % 10
            tmp.next = ListNode(sum_)
            tmp = tmp.next
        if carry > 0:
            tmp.next = ListNode(carry)
        return dummy_head.next




        # dummy_head = ListNode(0)
        # tmp = dummy_head
        # carry = 0
        # while (l1 is not None) & (l2 is not None):
        #     sum_ = l1.val + l2.val + carry
        #     if sum_ > 9:
        #         sum_ = sum_%10
        #         carry = 1
        #     else:
        #         carry = 0
        #     tmp.next = ListNode(sum_)
        #     tmp = tmp.next
        #     l1 = l1.next
        #     l2 = l2.next

        # if (l1 is None):
        #     while(l2 is not None):
        #         sum_ = l2.val + carry
        #         if sum_ > 9:
        #             sum_ = sum_%10
        #             carry = 1
        #         else:
        #             carry = 0
        #         tmp.next = ListNode(sum_)
        #         tmp = tmp.next
        #         l2 = l2.next
        #     if carry == 1:
        #         tmp.next = ListNode(1)
        # else:
        #     while(l1 is not None):
        #         sum_ = l1.val + carry
        #         if sum_ > 9:
        #             sum_ = sum_%10
        #             carry = 1
        #         else:
        #             carry = 0
        #         tmp.next = ListNode(sum_)
        #         tmp = tmp.next
        #         l1 = l1.next
        #     if carry == 1:
        #         tmp.next = ListNode(1)
        
        # return dummy_head.next

        # head = None
        # tmp = None
        # carry = 0
        # while (l1 is not None) or (l2 is not None):
        #     sum_ = carry
        #     if l1 is not None:
        #         sum_ = sum_ + l1.val
        #         l1 = l1.next
        #     if l2 is not None:
        #         sum_ = sum_ + l2.val
        #         l2 = l2.next
        #     carry = sum_//10
        #     sum_ = sum_%10

        #     node = ListNode(sum_)
        #     if tmp is None:
        #         tmp = node
        #         head = node
        #     else:
        #         tmp.next = node
        #         tmp = tmp.next

        # if carry :
        #     tmp.next = ListNode(carry)

        # return head

# @lc code=end

File: 2.add-two-numbers.cpp
/*
 * @lc app=leetcode id=2 lang=cpp
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head {new ListNode {0}};
        ListNode* cur {head};
        int offset {0};
        while(l1 || l2 || offset){
            if(l1){
                offset += l1->val;
                l1 = l1->next;
            }
            if(l2){
                offset += l2->val;
                l2 = l2->next;
            }
            cur->next = new ListNode {offset % 10};
            cur = cur->next;
            offset = offset / 10;
        }

        return head->next;


        // ListNode* head { new ListNode {0}};
        // ListNode* cur {head};
        // int offset {0};
        // while( l1 || l2){
        //     int num_one {0};
        //     int num_two {0};
        //     if (l1){
        //         num_one = l1->val;
        //         l1 = l1->next;
        //     }
        //     if (l2){
        //         num_two = l2->val;
        //         l2 = l2->next;
        //     }
        //     int sum_ = num_one + num_two + offset;
        //     offset = sum_ / 10;
        //     cur -> next = new ListNode {sum_ % 10};
        //     cur = cur->next;

        // }
        // if (offset > 0){
        //     cur->next = new ListNode {offset};
        // }
        // return head->next;
    



        
//         ListNode* head = new ListNode(0);
//         ListNode* cur = head;
//         int extra = 0;
//         while( l1 || l2 || extra){
//             if (l1){
//                 extra += l1->val;
//                 l1 = l1->next;
//             }
//             if (l2){
//                 extra += l2->val;
//                 l2 = l2->next;
//             }
//             cur->next = new ListNode(extra % 10);
//             extra = extra / 10;
//             cur = cur->next;
//         }
//         return head->next;
    }
};
// @lc code=end

